From 047b2ce8cb5cd4a0d29b2c863e684684d59d8006 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stephan=20D=C3=BCnner?= <stephan.duenner@bytesatwork.ch>
Date: Fri, 9 Aug 2019 19:17:33 +0200
Subject: [PATCH] cmd: Add cpuinfo
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Add command to print information about CPU.

Signed-off-by: Stephan Dünner <stephan.duenner@bytesatwork.ch>
Signed-off-by: Simone Schwizer <simone.schwizer@bytesatwork.ch>
(cherry picked from commit 28c5dbe6beaea21eb93c6ec844c4324207806e9c)
---
 MAINTAINERS   |  1 +
 cmd/Kconfig   |  6 ++++++
 cmd/Makefile  |  1 +
 cmd/cpuinfo.c | 37 +++++++++++++++++++++++++++++++++++++
 4 files changed, 45 insertions(+)
 create mode 100644 cmd/cpuinfo.c

diff --git a/MAINTAINERS b/MAINTAINERS
index a9f6672b652..fdd3adba649 100644
--- a/MAINTAINERS
+++ b/MAINTAINERS
@@ -272,6 +272,7 @@ S:	Maintained
 W:	https://www.bytesatwork.io/
 T:	git https://github.com/bytesatwork/u-boot-ti.git
 F:	.github/workflows/checkpatch.yml
+F:	cmd/cpuinfo.c
 
 ARM CORTINA ACCESS CAxxxx
 M:	Alex Nemirovsky <alex.nemirovsky@cortina-access.com>
diff --git a/cmd/Kconfig b/cmd/Kconfig
index 5c611fb3016..e8e02d6d30b 100644
--- a/cmd/Kconfig
+++ b/cmd/Kconfig
@@ -177,6 +177,12 @@ config CMD_CPU
 	  internal name) and clock frequency. Other information may be
 	  available depending on the CPU driver.
 
+config CMD_CPUINFO
+	bool "cpuinfo"
+	help
+	  Print information about the CPU such as type, revision, clock rate
+	  and temperature grade.
+
 config CMD_UFETCH
 	bool "U-Boot fetch"
 	help
diff --git a/cmd/Makefile b/cmd/Makefile
index 25479907797..f1a8b6e861a 100644
--- a/cmd/Makefile
+++ b/cmd/Makefile
@@ -52,6 +52,7 @@ obj-$(CONFIG_CMD_CONFIG) += config.o
 obj-$(CONFIG_CMD_CONITRACE) += conitrace.o
 obj-$(CONFIG_CMD_CONSOLE) += console.o
 obj-$(CONFIG_CMD_CPU) += cpu.o
+obj-$(CONFIG_CMD_CPUINFO) += cpuinfo.o
 obj-$(CONFIG_CMD_DATE) += date.o
 obj-$(CONFIG_CMD_DEMO) += demo.o
 obj-$(CONFIG_CMD_DM) += dm.o
diff --git a/cmd/cpuinfo.c b/cmd/cpuinfo.c
new file mode 100644
index 00000000000..9a5a0081f13
--- /dev/null
+++ b/cmd/cpuinfo.c
@@ -0,0 +1,37 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Copyright (C) 2025 bytesatwork AG - https://www.bytesatwork.io
+ */
+
+#include <command.h>
+#include <cpu.h>
+#include <dm.h>
+#include <init.h>
+
+static int do_cpuinfo(struct cmd_tbl *cmdtp, int flag, int argc, char * const argv[])
+{
+#if IS_ENABLED(CONFIG_DM) && IS_ENABLED(CONFIG_CPU)
+	struct udevice *dev;
+	char buf[256];
+	int ret;
+
+	ret = uclass_get_device(UCLASS_CPU, 0, &dev);
+	if (ret) {
+		printf("Failed to get CPU device: %d\n", ret);
+		return ret;
+	}
+
+	ret = cpu_get_desc(dev, buf, sizeof(buf));
+	if (ret) {
+		printf("Failed to get CPU description: %d\n", ret);
+		return ret;
+	}
+
+	puts(buf);
+	return 0;
+#else
+	return print_cpuinfo();
+#endif
+}
+
+U_BOOT_CMD(cpuinfo, 1, 1, do_cpuinfo, "show cpu info", "");
